### PR TITLE
1.14.0 and remove static builds

### PR DESCRIPTION
**Pull request by hmaarrfk**
I really don't think we should be dragging along the static builds. The build process isn't really nice toward building both the static and the dynamic library. I think we are going to be better off without this for 1.14.0

Let me know what you all think

- [ ] Check if I have to include static builds again https://github.com/conda-forge/hdf5-feedstock/issues/185
- [x] Discuss skipping the ROS3 test on windows -- https://github.com/HDFGroup/hdf5/issues/2406
- [x] Atomicity tests known to fail: https://github.com/HDFGroup/hdf5/issues/2196
- [x] Remove huge patch to rename executable files in cmake in favor of renaming them in the build script

Upstream changes

- Upstream changed from -show to -showconfig for cmake generated h5cc https://github.com/HDFGroup/hdf5/blob/e4fb9aa96ef805e2dada8a78c32aab8cf8c34f4a/release_docs/HISTORY-1_13.txt#L1111
- Patched out the use of the statically linked binaries for windows making the comrpessed package 3x smaller

TODO:
- [ ] Create a branch for 1.12.X maintenance (this will be necessary since HDF5 is typically a hard transition)

Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
The really heavy handed patching is to remove all the statically linked binaries from windows:
image

**Comment by hmaarrfk** [comment]
The executables that link to the shared libraries take 10s of kB, those that statically link take MBs
image

**Comment by hmaarrfk** [comment]
@djhoese i'm not sure if you are still interested in ROS3 functionality on windows.

**Comment by djhoese** [comment]
It isn't clear to me what ROS3 has to do with this PR or static builds. ROS3 should be made to work on all platforms if upstream supports it. If the tests fail on Windows due to some CI limitation then OK.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -81,127 +90,84 @@ requirements:
...
+    {% for each_hdf5_lib in hdf5_libs %}
+    - test -f $PREFIX/lib/lib{{ each_hdf5_lib }}${SHLIB_EXT}  # [not win]
+    - if not exist %PREFIX%\\Library\\lib\\{{ each_hdf5_lib }}.lib exit 1  # [win]
+    - if not exist %PREFIX%\\Library\\bin\\{{ each_hdf5_lib }}.dll exit 1  # [win]
+    {% endfor %}
+    # Check ros3 driver
+    - h5dump --filedriver=ros3 "http://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | grep '^HDF5'  # [unix]
+    # Jan 2023 -- hmaarrfk
+    # I am really not sure why the ROS driver test doesn't pass on windows
+    # - h5dump --filedriver=ros3 "https://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | findstr '^HDF5'  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:170]: @djhoese i had to skip the windows test here.

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

File: recipe/meta.yaml (hunk 1 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:168]: Trying it locally, i also get 
```
h5dump error: unable to create FAPL for file access
```

Comment (hmaarrfk) [reply recipe/meta.yaml:168]: Yeah, i'm not sure, on windows the header has the line uncommented meaning that cmake thinks that it compiled with ROS3 support
```
#define H5_HAVE_ROS3_VFD 1
```

Comment (djhoese) [reply recipe/meta.yaml:168]: I feel like I might be missing some context here. Regardless, is this something that should be filed as a bug/discussion upstream?

Comment (hmaarrfk) [reply recipe/meta.yaml:168]: done! they are on it! See link at the top level post!

**Comment by mkitti** [comment]
Do we need to synchronize this with a h5py update?

h5py 3.8.0 was released a few hours ago:
https://github.com/h5py/h5py/releases/tag/3.8.0

**Comment by mkitti** [comment]
h5py was updated to 3.8.0:
https://github.com/conda-forge/h5py-feedstock/pull/125

This incorporates the patches needed for 1.14 support.

**Comment by hmaarrfk** [comment]
> Do we need to synchronize this with a h5py update?

In essence, not really. h5py is pinned to the previous version of HDF5 through the run_export.

So this could be merged, without any fear of breaking existing user's H5PY installation.

Im mostly leaving this here for people to comment on the removal of the static builds. But to me, it seems that they were added as an abundance of caution.

**Comment by mkitti** [comment]
It would be nice if the shared versions were aliased to the static versions in this case via a symlink or just copying so that typing `h5ls` would still work.

**Comment by hmaarrfk** [comment]
`h5ls-dynamic` seems to be a creation of the `cmake` files.

doesn't h5ls still work? The tests say they do.

I aggressively patched windows to make `h5ls` exist but to remove `h5ls-dynamic`. I'm honestly now proud of the aggressive patching. Maybe we should just "rename" the executable instead of patching so much. I'm happy to report this as an issue upstream as well.

**Comment by mkitti** [comment]
Ah, I see. It might be worth discussing at https://forum.hdfgroup.org/ first?

**Comment by mkitti** [comment]
It might make sense to split the tools off to another package. Some Linux distributions do this:
https://packages.debian.org/bullseye/hdf5-tools
https://pkgs.org/search/?q=hdf5-tools

**Comment by hmaarrfk** [comment]
> discussing at https://forum.hdfgroup.org/ first?

Presently, I am too tired to articulate to a wider audience, but maybe over the weekend i can formulate my thoughts?

I am tempted to reimplement this as a "file rename" in our build script instead of a huge patch.

**Comment by hmaarrfk** [comment]
> It might make sense to split the tools off to another package

Two aspects: 

1. conda doesn't really make this "easy". For example, with RPMs I can specify a list of files I want in a subpackage, then everything else goes in the `main` package. This ensures I don't "miss" anything in the subpackage creation. This process is a little manual, and error prone.
2. conda has historically been more "batteries included" (including more options than "the bare minimum") than other distributions.  The binaries on linux only take up 2MB / 11.1MB as reported by `ncdu`. Unless somebody wants to step up and "help maintain", I'm not so inclined to split the package given `1.`.  As a note, uncompressed, the headers make up 2.8/11.1 MB.

**Comment by hmaarrfk** [comment]
@conda-forge-admin please restart cis

**Comment by hmaarrfk** [comment]
@conda-forge/hdf5 this is ready for review.

I'm hoping to merge Feb 5.

**Review by ocefpaf — APPROVED** [review]

File: recipe/meta.yaml [hunk 2]
```diff
@@ -75,133 +80,91 @@ requirements:
...
+    {% for each_hdf5_lib in hdf5_libs %}
+    - test -f $PREFIX/lib/lib{{ each_hdf5_lib }}${SHLIB_EXT}  # [not win]
+    - if not exist %PREFIX%\\Library\\lib\\{{ each_hdf5_lib }}.lib exit 1  # [win]
+    - if not exist %PREFIX%\\Library\\bin\\{{ each_hdf5_lib }}.dll exit 1  # [win]
+    {% endfor %}
+    # Check ros3 driver
+    - h5dump --filedriver=ros3 "http://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | grep '^HDF5'  # [unix]
+    # Jan 2023 -- hmaarrfk
+    # I am really not sure why the ROS driver test doesn't pass on windows
+    - h5dump --filedriver=ros3 --s3-cred="(,,)" "https://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5"  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:167]:
```suggestion
    # Jan 2023 -- hmaarrfk. Tests will fail on windows without --s3-cred specified
    # https://github.com/HDFGroup/hdf5/issues/2406
    - h5dump --filedriver=ros3 --s3-cred="(,,)" "https://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5"  # [win]
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -75,133 +80,91 @@ requirements:
...
+    {% for each_hdf5_lib in hdf5_libs %}
+    - test -f $PREFIX/lib/lib{{ each_hdf5_lib }}${SHLIB_EXT}  # [not win]
+    - if not exist %PREFIX%\\Library\\lib\\{{ each_hdf5_lib }}.lib exit 1  # [win]
+    - if not exist %PREFIX%\\Library\\bin\\{{ each_hdf5_lib }}.dll exit 1  # [win]
+    {% endfor %}
+    # Check ros3 driver
+    - h5dump --filedriver=ros3 "http://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | grep '^HDF5'  # [unix]
+    # Jan 2023 -- hmaarrfk. Tests will fail on windows without --s3-cred specified
+    # https://github.com/HDFGroup/hdf5/issues/2406
+    - h5dump --filedriver=ros3 --s3-cred="(,,)" "https://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | findstr '^HDF5'  # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:167]:
```suggestion
    - h5dump --filedriver=ros3 --s3-cred="(,,)" "https://s3.amazonaws.com/hdfgroup/data/hdf5demo/tall.h5" | findstr "^HDF5"  # [win]
```

**Comment by hmaarrfk** [comment]
Thank you all for your reviews.

**Pull request merged**